<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="The plotting functions">

<title>core – geneorder</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-fe8f458f31a467b51aeabd2275c1e88a.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="core – geneorder">
<meta property="og:description" content="The plotting functions">
<meta property="og:image" content="https://galicae.github.io/geneorder/00_core_files/figure-html/cell-4-output-1.png">
<meta property="og:site_name" content="geneorder">
<meta property="og:image:height" content="261">
<meta property="og:image:width" content="1120">
<meta name="twitter:title" content="core – geneorder">
<meta name="twitter:description" content="The plotting functions">
<meta name="twitter:image" content="https://galicae.github.io/geneorder/00_core_files/figure-html/cell-4-output-1.png">
<meta name="twitter:image-height" content="261">
<meta name="twitter:image-width" content="1120">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">geneorder</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./core.html">core</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">geneorder</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./core.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">core</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./util.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">util</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#plot_synteny" id="toc-plot_synteny" class="nav-link active" data-scroll-target="#plot_synteny">plot_synteny</a></li>
  <li><a href="#plot_synteny_schematic" id="toc-plot_synteny_schematic" class="nav-link" data-scroll-target="#plot_synteny_schematic">plot_synteny_schematic</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/galicae/geneorder/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div><div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="core.html.md"><i class="bi bi-file-code"></i>CommonMark</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">core</h1>
</div>

<div>
  <div class="description">
    The plotting functions
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p>Before we get started with plotting, we should read in some toy data. This is a modified excerpt from the <em>P. litorale</em> genome annotation, specifically the part that holds the Hox gene cluster.</p>
<div id="cell-3" class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>gff <span class="op">=</span> util.read_gff(os.environ[<span class="st">"EXAMPLE_DATA_PATH"</span>] <span class="op">+</span> <span class="st">"plit.gff3"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>util.decorate(gff, attributes<span class="op">=</span>{<span class="st">"gene_id"</span>: <span class="st">"ID"</span>, <span class="st">"gene_name"</span>: <span class="st">"gene"</span>})</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>hox_genes <span class="op">=</span> [</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"PB.8615"</span>,</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"g9718"</span>,</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"PB.8616"</span>,</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"g9720"</span>,</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"g9721"</span>,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"PB.8617"</span>,</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"g9723"</span>,</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"g9724"</span>,</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"g9725"</span>,</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>keep <span class="op">=</span> gff[<span class="st">"gene_id"</span>].isin(hox_genes)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>hox <span class="op">=</span> gff[keep].reset_index(drop<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<p><a href="https://github.com/galicae/geneorder/blob/main/geneorder/core.py#L21" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="plot_synteny" class="level3">
<h3 class="anchored" data-anchor-id="plot_synteny">plot_synteny</h3>
<blockquote class="blockquote">
<pre><code> plot_synteny (gff, block_start:int=None, block_end:int=None,
               figsize:(&lt;class'float'&gt;,&lt;class'float'&gt;)=None,
               locus_start:str='start', locus_end:str='end',
               molecule:str='seqid', locus_name:str='gene_name',
               fig_title:str=None, save:str|pathlib.Path='synteny.svg',
               palette:dict|None={'aliceblue': '#F0F8FF', 'antiquewhite':
               '#FAEBD7', 'aqua': '#00FFFF', 'aquamarine': '#7FFFD4',
               'azure': '#F0FFFF', 'beige': '#F5F5DC', 'bisque':
               '#FFE4C4', 'black': '#000000', 'blanchedalmond': '#FFEBCD',
               'blue': '#0000FF', 'blueviolet': '#8A2BE2', 'brown':
               '#A52A2A', 'burlywood': '#DEB887', 'cadetblue': '#5F9EA0',
               'chartreuse': '#7FFF00', 'chocolate': '#D2691E', 'coral':
               '#FF7F50', 'cornflowerblue': '#6495ED', 'cornsilk':
               '#FFF8DC', 'crimson': '#DC143C', 'cyan': '#00FFFF',
               'darkblue': '#00008B', 'darkcyan': '#008B8B',
               'darkgoldenrod': '#B8860B', 'darkgray': '#A9A9A9',
               'darkgreen': '#006400', 'darkgrey': '#A9A9A9', 'darkkhaki':
               '#BDB76B', 'darkmagenta': '#8B008B', 'darkolivegreen':
               '#556B2F', 'darkorange': '#FF8C00', 'darkorchid':
               '#9932CC', 'darkred': '#8B0000', 'darksalmon': '#E9967A',
               'darkseagreen': '#8FBC8F', 'darkslateblue': '#483D8B',
               'darkslategray': '#2F4F4F', 'darkslategrey': '#2F4F4F',
               'darkturquoise': '#00CED1', 'darkviolet': '#9400D3',
               'deeppink': '#FF1493', 'deepskyblue': '#00BFFF', 'dimgray':
               '#696969', 'dimgrey': '#696969', 'dodgerblue': '#1E90FF',
               'firebrick': '#B22222', 'floralwhite': '#FFFAF0',
               'forestgreen': '#228B22', 'fuchsia': '#FF00FF',
               'gainsboro': '#DCDCDC', 'ghostwhite': '#F8F8FF', 'gold':
               '#FFD700', 'goldenrod': '#DAA520', 'gray': '#808080',
               'green': '#008000', 'greenyellow': '#ADFF2F', 'grey':
               '#808080', 'honeydew': '#F0FFF0', 'hotpink': '#FF69B4',
               'indianred': '#CD5C5C', 'indigo': '#4B0082', 'ivory':
               '#FFFFF0', 'khaki': '#F0E68C', 'lavender': '#E6E6FA',
               'lavenderblush': '#FFF0F5', 'lawngreen': '#7CFC00',
               'lemonchiffon': '#FFFACD', 'lightblue': '#ADD8E6',
               'lightcoral': '#F08080', 'lightcyan': '#E0FFFF',
               'lightgoldenrodyellow': '#FAFAD2', 'lightgray': '#D3D3D3',
               'lightgreen': '#90EE90', 'lightgrey': '#D3D3D3',
               'lightpink': '#FFB6C1', 'lightsalmon': '#FFA07A',
               'lightseagreen': '#20B2AA', 'lightskyblue': '#87CEFA',
               'lightslategray': '#778899', 'lightslategrey': '#778899',
               'lightsteelblue': '#B0C4DE', 'lightyellow': '#FFFFE0',
               'lime': '#00FF00', 'limegreen': '#32CD32', 'linen':
               '#FAF0E6', 'magenta': '#FF00FF', 'maroon': '#800000',
               'mediumaquamarine': '#66CDAA', 'mediumblue': '#0000CD',
               'mediumorchid': '#BA55D3', 'mediumpurple': '#9370DB',
               'mediumseagreen': '#3CB371', 'mediumslateblue': '#7B68EE',
               'mediumspringgreen': '#00FA9A', 'mediumturquoise':
               '#48D1CC', 'mediumvioletred': '#C71585', 'midnightblue':
               '#191970', 'mintcream': '#F5FFFA', 'mistyrose': '#FFE4E1',
               'moccasin': '#FFE4B5', 'navajowhite': '#FFDEAD', 'navy':
               '#000080', 'oldlace': '#FDF5E6', 'olive': '#808000',
               'olivedrab': '#6B8E23', 'orange': '#FFA500', 'orangered':
               '#FF4500', 'orchid': '#DA70D6', 'palegoldenrod': '#EEE8AA',
               'palegreen': '#98FB98', 'paleturquoise': '#AFEEEE',
               'palevioletred': '#DB7093', 'papayawhip': '#FFEFD5',
               'peachpuff': '#FFDAB9', 'peru': '#CD853F', 'pink':
               '#FFC0CB', 'plum': '#DDA0DD', 'powderblue': '#B0E0E6',
               'purple': '#800080', 'rebeccapurple': '#663399', 'red':
               '#FF0000', 'rosybrown': '#BC8F8F', 'royalblue': '#4169E1',
               'saddlebrown': '#8B4513', 'salmon': '#FA8072',
               'sandybrown': '#F4A460', 'seagreen': '#2E8B57', 'seashell':
               '#FFF5EE', 'sienna': '#A0522D', 'silver': '#C0C0C0',
               'skyblue': '#87CEEB', 'slateblue': '#6A5ACD', 'slategray':
               '#708090', 'slategrey': '#708090', 'snow': '#FFFAFA',
               'springgreen': '#00FF7F', 'steelblue': '#4682B4', 'tan':
               '#D2B48C', 'teal': '#008080', 'thistle': '#D8BFD8',
               'tomato': '#FF6347', 'turquoise': '#40E0D0', 'violet':
               '#EE82EE', 'wheat': '#F5DEB3', 'white': '#FFFFFF',
               'whitesmoke': '#F5F5F5', 'yellow': '#FFFF00',
               'yellowgreen': '#9ACD32'},
               chromosome_color:str='lightgray',
               backup_gene_color:str='darkgray',
               gene_name_offset:float=0.02, chromosome_width:float=1,
               gene_width:float=5)</code></pre>
</blockquote>
<p><em>Function to plot a syntenic cluster of genes.</em></p>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>gff</td>
<td></td>
<td></td>
<td>a GFF in Pandas dataframe form. Only includes the genes of the syntenic block in question.</td>
</tr>
<tr class="even">
<td>block_start</td>
<td>int</td>
<td>None</td>
<td>The start coordinate of the syntenic locus to plot. If None, will be set to the start of the first gene minus 5% of the block length.</td>
</tr>
<tr class="odd">
<td>block_end</td>
<td>int</td>
<td>None</td>
<td>The end coordinate of the syntenic locus to plot. If None, will be set to the end of the last gene plus 5% of the block length.</td>
</tr>
<tr class="even">
<td>figsize</td>
<td>(&lt;class ‘float’&gt;, &lt;class ‘float’&gt;)</td>
<td>None</td>
<td>the figure size.</td>
</tr>
<tr class="odd">
<td>locus_start</td>
<td>str</td>
<td>start</td>
<td>the GFF column that describes the start coordinate of the entities (e.g.&nbsp;gene, mRNA) to be plotted on the chromosome. In a well-behaved GFF, this should be “start”.</td>
</tr>
<tr class="even">
<td>locus_end</td>
<td>str</td>
<td>end</td>
<td>the GFF column that describes the end coordinate of the entities (e.g.&nbsp;gene, mRNA) to be plotted on the chromosome. In a well-behaved GFF, this should be “end”.</td>
</tr>
<tr class="odd">
<td>molecule</td>
<td>str</td>
<td>seqid</td>
<td>the GFF column that holds the molecule name (chromosome/scaffold/contig ID). In a well-behaved GFF, this should be “seqid”.</td>
</tr>
<tr class="even">
<td>locus_name</td>
<td>str</td>
<td>gene_name</td>
<td>the GFF column that holds the gene/mRNA symbol. This is usually a tag in the “attributes” column of a well-behaved GFF and should have been extracted prior to this step. It will be plotted according to the <code>gene_name_offset</code> parameter.</td>
</tr>
<tr class="odd">
<td>fig_title</td>
<td>str</td>
<td>None</td>
<td>Figure title.</td>
</tr>
<tr class="even">
<td>save</td>
<td>str | pathlib.Path</td>
<td>synteny.svg</td>
<td>path to save an SVG image.</td>
</tr>
<tr class="odd">
<td>palette</td>
<td>dict | None</td>
<td>{‘aliceblue’: ‘#F0F8FF’, ‘antiquewhite’: ‘#FAEBD7’, ‘aqua’: ‘#00FFFF’, ‘aquamarine’: ‘#7FFFD4’, ‘azure’: ‘#F0FFFF’, ‘beige’: ‘#F5F5DC’, ‘bisque’: ‘#FFE4C4’, ‘black’: ‘#000000’, ‘blanchedalmond’: ‘#FFEBCD’, ‘blue’: ‘#0000FF’, ‘blueviolet’: ‘#8A2BE2’, ‘brown’: ‘#A52A2A’, ‘burlywood’: ‘#DEB887’, ‘cadetblue’: ‘#5F9EA0’, ‘chartreuse’: ‘#7FFF00’, ‘chocolate’: ‘#D2691E’, ‘coral’: ‘#FF7F50’, ‘cornflowerblue’: ‘#6495ED’, ‘cornsilk’: ‘#FFF8DC’, ‘crimson’: ‘#DC143C’, ‘cyan’: ‘#00FFFF’, ‘darkblue’: ‘#00008B’, ‘darkcyan’: ‘#008B8B’, ‘darkgoldenrod’: ‘#B8860B’, ‘darkgray’: ‘#A9A9A9’, ‘darkgreen’: ‘#006400’, ‘darkgrey’: ‘#A9A9A9’, ‘darkkhaki’: ‘#BDB76B’, ‘darkmagenta’: ‘#8B008B’, ‘darkolivegreen’: ‘#556B2F’, ‘darkorange’: ‘#FF8C00’, ‘darkorchid’: ‘#9932CC’, ‘darkred’: ‘#8B0000’, ‘darksalmon’: ‘#E9967A’, ‘darkseagreen’: ‘#8FBC8F’, ‘darkslateblue’: ‘#483D8B’, ‘darkslategray’: ‘#2F4F4F’, ‘darkslategrey’: ‘#2F4F4F’, ‘darkturquoise’: ‘#00CED1’, ‘darkviolet’: ‘#9400D3’, ‘deeppink’: ‘#FF1493’, ‘deepskyblue’: ‘#00BFFF’, ‘dimgray’: ‘#696969’, ‘dimgrey’: ‘#696969’, ‘dodgerblue’: ‘#1E90FF’, ‘firebrick’: ‘#B22222’, ‘floralwhite’: ‘#FFFAF0’, ‘forestgreen’: ‘#228B22’, ‘fuchsia’: ‘#FF00FF’, ‘gainsboro’: ‘#DCDCDC’, ‘ghostwhite’: ‘#F8F8FF’, ‘gold’: ‘#FFD700’, ‘goldenrod’: ‘#DAA520’, ‘gray’: ‘#808080’, ‘green’: ‘#008000’, ‘greenyellow’: ‘#ADFF2F’, ‘grey’: ‘#808080’, ‘honeydew’: ‘#F0FFF0’, ‘hotpink’: ‘#FF69B4’, ‘indianred’: ‘#CD5C5C’, ‘indigo’: ‘#4B0082’, ‘ivory’: ‘#FFFFF0’, ‘khaki’: ‘#F0E68C’, ‘lavender’: ‘#E6E6FA’, ‘lavenderblush’: ‘#FFF0F5’, ‘lawngreen’: ‘#7CFC00’, ‘lemonchiffon’: ‘#FFFACD’, ‘lightblue’: ‘#ADD8E6’, ‘lightcoral’: ‘#F08080’, ‘lightcyan’: ‘#E0FFFF’, ‘lightgoldenrodyellow’: ‘#FAFAD2’, ‘lightgray’: ‘#D3D3D3’, ‘lightgreen’: ‘#90EE90’, ‘lightgrey’: ‘#D3D3D3’, ‘lightpink’: ‘#FFB6C1’, ‘lightsalmon’: ‘#FFA07A’, ‘lightseagreen’: ‘#20B2AA’, ‘lightskyblue’: ‘#87CEFA’, ‘lightslategray’: ‘#778899’, ‘lightslategrey’: ‘#778899’, ‘lightsteelblue’: ‘#B0C4DE’, ‘lightyellow’: ‘#FFFFE0’, ‘lime’: ‘#00FF00’, ‘limegreen’: ‘#32CD32’, ‘linen’: ‘#FAF0E6’, ‘magenta’: ‘#FF00FF’, ‘maroon’: ‘#800000’, ‘mediumaquamarine’: ‘#66CDAA’, ‘mediumblue’: ‘#0000CD’, ‘mediumorchid’: ‘#BA55D3’, ‘mediumpurple’: ‘#9370DB’, ‘mediumseagreen’: ‘#3CB371’, ‘mediumslateblue’: ‘#7B68EE’, ‘mediumspringgreen’: ‘#00FA9A’, ‘mediumturquoise’: ‘#48D1CC’, ‘mediumvioletred’: ‘#C71585’, ‘midnightblue’: ‘#191970’, ‘mintcream’: ‘#F5FFFA’, ‘mistyrose’: ‘#FFE4E1’, ‘moccasin’: ‘#FFE4B5’, ‘navajowhite’: ‘#FFDEAD’, ‘navy’: ‘#000080’, ‘oldlace’: ‘#FDF5E6’, ‘olive’: ‘#808000’, ‘olivedrab’: ‘#6B8E23’, ‘orange’: ‘#FFA500’, ‘orangered’: ‘#FF4500’, ‘orchid’: ‘#DA70D6’, ‘palegoldenrod’: ‘#EEE8AA’, ‘palegreen’: ‘#98FB98’, ‘paleturquoise’: ‘#AFEEEE’, ‘palevioletred’: ‘#DB7093’, ‘papayawhip’: ‘#FFEFD5’, ‘peachpuff’: ‘#FFDAB9’, ‘peru’: ‘#CD853F’, ‘pink’: ‘#FFC0CB’, ‘plum’: ‘#DDA0DD’, ‘powderblue’: ‘#B0E0E6’, ‘purple’: ‘#800080’, ‘rebeccapurple’: ‘#663399’, ‘red’: ‘#FF0000’, ‘rosybrown’: ‘#BC8F8F’, ‘royalblue’: ‘#4169E1’, ‘saddlebrown’: ‘#8B4513’, ‘salmon’: ‘#FA8072’, ‘sandybrown’: ‘#F4A460’, ‘seagreen’: ‘#2E8B57’, ‘seashell’: ‘#FFF5EE’, ‘sienna’: ‘#A0522D’, ‘silver’: ‘#C0C0C0’, ‘skyblue’: ‘#87CEEB’, ‘slateblue’: ‘#6A5ACD’, ‘slategray’: ‘#708090’, ‘slategrey’: ‘#708090’, ‘snow’: ‘#FFFAFA’, ‘springgreen’: ‘#00FF7F’, ‘steelblue’: ‘#4682B4’, ‘tan’: ‘#D2B48C’, ‘teal’: ‘#008080’, ‘thistle’: ‘#D8BFD8’, ‘tomato’: ‘#FF6347’, ‘turquoise’: ‘#40E0D0’, ‘violet’: ‘#EE82EE’, ‘wheat’: ‘#F5DEB3’, ‘white’: ‘#FFFFFF’, ‘whitesmoke’: ‘#F5F5F5’, ‘yellow’: ‘#FFFF00’, ‘yellowgreen’: ‘#9ACD32’}</td>
<td></td>
</tr>
<tr class="even">
<td>chromosome_color</td>
<td>str</td>
<td>lightgray</td>
<td>the color of the line representing the molecule the genes are plotted on.</td>
</tr>
<tr class="odd">
<td>backup_gene_color</td>
<td>str</td>
<td>darkgray</td>
<td>If the GFF has no column titled “color”, the gene arrows will be filled using this color.</td>
</tr>
<tr class="even">
<td>gene_name_offset</td>
<td>float</td>
<td>0.02</td>
<td>Default offset for plotting gene names if the GFF has no column titled “offset”.</td>
</tr>
<tr class="odd">
<td>chromosome_width</td>
<td>float</td>
<td>1</td>
<td>Linewidth parameter for the line representing the chromosome.</td>
</tr>
<tr class="even">
<td>gene_width</td>
<td>float</td>
<td>5</td>
<td>Width of the arrow patches representing genes. Please refer to the documentation of <code>matplotlib.pyplot.arrow</code> for more details.</td>
</tr>
<tr class="odd">
<td><strong>Returns</strong></td>
<td><strong>None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>We can use this function with any GFF dataframe that is reduced to one row per gene (for example a <code>type: gene</code> or <code>type: mRNA</code> entry per gene).</p>
<div id="cell-6" class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>plot_synteny(hox, figsize<span class="op">=</span>(<span class="dv">11</span>, <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-4-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>If the input GFF contains information about the color and the gene name offsets, the plotter will use them:</p>
<div id="cell-8" class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"color"</span>] <span class="op">=</span> [</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"red"</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"orange"</span>,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gold"</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"lightgreen"</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"forestgreen"</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"royalblue"</span>,</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"darkblue"</span>,</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"darkmagenta"</span>,</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"magenta"</span>,</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"offset"</span>] <span class="op">=</span> [<span class="fl">0.02</span>, <span class="op">-</span><span class="fl">0.025</span>, <span class="fl">0.02</span>, <span class="fl">0.02</span>, <span class="fl">0.02</span>, <span class="op">-</span><span class="fl">0.025</span>, <span class="fl">0.02</span>, <span class="fl">0.02</span>, <span class="fl">0.02</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-9" class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>plot_synteny(hox, figsize<span class="op">=</span>(<span class="dv">11</span>, <span class="dv">2</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"strand"</span>] <span class="op">=</span> [<span class="st">"+"</span>, <span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"-"</span>, <span class="st">"-"</span>, <span class="st">"-"</span>, <span class="st">"+"</span>, <span class="st">"+"</span>, <span class="st">"+"</span>]</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>plot_synteny(hox, figsize<span class="op">=</span>(<span class="dv">11</span>, <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-6-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-6-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can also use hex codes directly for coloring:</p>
<div id="cell-11" class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"color"</span>] <span class="op">=</span> [</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#E7DAD2"</span>,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#CFE2B7"</span>,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#FA7F6F"</span>,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#83B0D2"</span>,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#BEB9DC"</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#C0E59A"</span>,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#8ECFC9"</span>,</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#FFBE79"</span>,</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"#BAAC8C"</span>,</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-12" class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>plot_synteny(hox, figsize<span class="op">=</span>(<span class="dv">11</span>, <span class="dv">2</span>), palette<span class="op">=</span><span class="va">None</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-8-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
<p><a href="https://github.com/galicae/geneorder/blob/main/geneorder/core.py#L121" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="plot_synteny_schematic" class="level3">
<h3 class="anchored" data-anchor-id="plot_synteny_schematic">plot_synteny_schematic</h3>
<blockquote class="blockquote">
<pre><code> plot_synteny_schematic (gff:pandas.core.frame.DataFrame,
                         block_gene:float=400, block_dist:float=200,
                         locus_start:str='start', locus_end:str='end',
                         molecule:str='seqid', locus_name:str='gene_name',
                         interrupted:bool=False,
                         palette:dict|None={'aliceblue': '#F0F8FF',
                         'antiquewhite': '#FAEBD7', 'aqua': '#00FFFF',
                         'aquamarine': '#7FFFD4', 'azure': '#F0FFFF',
                         'beige': '#F5F5DC', 'bisque': '#FFE4C4', 'black':
                         '#000000', 'blanchedalmond': '#FFEBCD', 'blue':
                         '#0000FF', 'blueviolet': '#8A2BE2', 'brown':
                         '#A52A2A', 'burlywood': '#DEB887', 'cadetblue':
                         '#5F9EA0', 'chartreuse': '#7FFF00', 'chocolate':
                         '#D2691E', 'coral': '#FF7F50', 'cornflowerblue':
                         '#6495ED', 'cornsilk': '#FFF8DC', 'crimson':
                         '#DC143C', 'cyan': '#00FFFF', 'darkblue':
                         '#00008B', 'darkcyan': '#008B8B',
                         'darkgoldenrod': '#B8860B', 'darkgray':
                         '#A9A9A9', 'darkgreen': '#006400', 'darkgrey':
                         '#A9A9A9', 'darkkhaki': '#BDB76B', 'darkmagenta':
                         '#8B008B', 'darkolivegreen': '#556B2F',
                         'darkorange': '#FF8C00', 'darkorchid': '#9932CC',
                         'darkred': '#8B0000', 'darksalmon': '#E9967A',
                         'darkseagreen': '#8FBC8F', 'darkslateblue':
                         '#483D8B', 'darkslategray': '#2F4F4F',
                         'darkslategrey': '#2F4F4F', 'darkturquoise':
                         '#00CED1', 'darkviolet': '#9400D3', 'deeppink':
                         '#FF1493', 'deepskyblue': '#00BFFF', 'dimgray':
                         '#696969', 'dimgrey': '#696969', 'dodgerblue':
                         '#1E90FF', 'firebrick': '#B22222', 'floralwhite':
                         '#FFFAF0', 'forestgreen': '#228B22', 'fuchsia':
                         '#FF00FF', 'gainsboro': '#DCDCDC', 'ghostwhite':
                         '#F8F8FF', 'gold': '#FFD700', 'goldenrod':
                         '#DAA520', 'gray': '#808080', 'green': '#008000',
                         'greenyellow': '#ADFF2F', 'grey': '#808080',
                         'honeydew': '#F0FFF0', 'hotpink': '#FF69B4',
                         'indianred': '#CD5C5C', 'indigo': '#4B0082',
                         'ivory': '#FFFFF0', 'khaki': '#F0E68C',
                         'lavender': '#E6E6FA', 'lavenderblush':
                         '#FFF0F5', 'lawngreen': '#7CFC00',
                         'lemonchiffon': '#FFFACD', 'lightblue':
                         '#ADD8E6', 'lightcoral': '#F08080', 'lightcyan':
                         '#E0FFFF', 'lightgoldenrodyellow': '#FAFAD2',
                         'lightgray': '#D3D3D3', 'lightgreen': '#90EE90',
                         'lightgrey': '#D3D3D3', 'lightpink': '#FFB6C1',
                         'lightsalmon': '#FFA07A', 'lightseagreen':
                         '#20B2AA', 'lightskyblue': '#87CEFA',
                         'lightslategray': '#778899', 'lightslategrey':
                         '#778899', 'lightsteelblue': '#B0C4DE',
                         'lightyellow': '#FFFFE0', 'lime': '#00FF00',
                         'limegreen': '#32CD32', 'linen': '#FAF0E6',
                         'magenta': '#FF00FF', 'maroon': '#800000',
                         'mediumaquamarine': '#66CDAA', 'mediumblue':
                         '#0000CD', 'mediumorchid': '#BA55D3',
                         'mediumpurple': '#9370DB', 'mediumseagreen':
                         '#3CB371', 'mediumslateblue': '#7B68EE',
                         'mediumspringgreen': '#00FA9A',
                         'mediumturquoise': '#48D1CC', 'mediumvioletred':
                         '#C71585', 'midnightblue': '#191970',
                         'mintcream': '#F5FFFA', 'mistyrose': '#FFE4E1',
                         'moccasin': '#FFE4B5', 'navajowhite': '#FFDEAD',
                         'navy': '#000080', 'oldlace': '#FDF5E6', 'olive':
                         '#808000', 'olivedrab': '#6B8E23', 'orange':
                         '#FFA500', 'orangered': '#FF4500', 'orchid':
                         '#DA70D6', 'palegoldenrod': '#EEE8AA',
                         'palegreen': '#98FB98', 'paleturquoise':
                         '#AFEEEE', 'palevioletred': '#DB7093',
                         'papayawhip': '#FFEFD5', 'peachpuff': '#FFDAB9',
                         'peru': '#CD853F', 'pink': '#FFC0CB', 'plum':
                         '#DDA0DD', 'powderblue': '#B0E0E6', 'purple':
                         '#800080', 'rebeccapurple': '#663399', 'red':
                         '#FF0000', 'rosybrown': '#BC8F8F', 'royalblue':
                         '#4169E1', 'saddlebrown': '#8B4513', 'salmon':
                         '#FA8072', 'sandybrown': '#F4A460', 'seagreen':
                         '#2E8B57', 'seashell': '#FFF5EE', 'sienna':
                         '#A0522D', 'silver': '#C0C0C0', 'skyblue':
                         '#87CEEB', 'slateblue': '#6A5ACD', 'slategray':
                         '#708090', 'slategrey': '#708090', 'snow':
                         '#FFFAFA', 'springgreen': '#00FF7F', 'steelblue':
                         '#4682B4', 'tan': '#D2B48C', 'teal': '#008080',
                         'thistle': '#D8BFD8', 'tomato': '#FF6347',
                         'turquoise': '#40E0D0', 'violet': '#EE82EE',
                         'wheat': '#F5DEB3', 'white': '#FFFFFF',
                         'whitesmoke': '#F5F5F5', 'yellow': '#FFFF00',
                         'yellowgreen': '#9ACD32'}, fontsize:int=30,
                         chromosome_color:str='black',
                         backup_gene_color:str='darkgray',
                         gene_name_offset:float=0.02,
                         chromosome_width:float=3, head_length:float=200,
                         head_width:float=1.7, gene_width:float=1,
                         ax:matplotlib.axes._axes.Axes=None,
                         figsize:(&lt;class'float'&gt;,&lt;class'float'&gt;)=None,
                         save:str|pathlib.Path='schematic.svg',
                         return_fig:bool=False)</code></pre>
</blockquote>
<table class="caption-top table">
<colgroup>
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 34%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Default</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>gff</td>
<td>DataFrame</td>
<td></td>
<td>a GFF in Pandas dataframe form. Only includes the genes of the syntenic block in question. Assumed to be sorted in plotting order.</td>
</tr>
<tr class="even">
<td>block_gene</td>
<td>float</td>
<td>400</td>
<td>length of a gene arrow, in plot coordinate space.</td>
</tr>
<tr class="odd">
<td>block_dist</td>
<td>float</td>
<td>200</td>
<td>length of gap between successive genes, in plot coordinate space.</td>
</tr>
<tr class="even">
<td>locus_start</td>
<td>str</td>
<td>start</td>
<td>the GFF column that describes the start coordinate of the entities (e.g.&nbsp;gene, mRNA) to be plotted on the chromosome. In a well-behaved GFF, this should be “start”.</td>
</tr>
<tr class="odd">
<td>locus_end</td>
<td>str</td>
<td>end</td>
<td>the GFF column that describes the end coordinate of the entities (e.g.&nbsp;gene, mRNA) to be plotted on the chromosome. In a well-behaved GFF, this should be “end”.</td>
</tr>
<tr class="even">
<td>molecule</td>
<td>str</td>
<td>seqid</td>
<td>the GFF column that holds the molecule name (chromosome/scaffold/contig ID). In a well-behaved GFF, this should be “seqid”.</td>
</tr>
<tr class="odd">
<td>locus_name</td>
<td>str</td>
<td>gene_name</td>
<td>the GFF column that holds the gene/mRNA symbol. This is usually a tag in the “attributes” column of a well-behaved GFF and should have been extracted prior to this step. It will be plotted according to the <code>gene_name_offset</code> parameter.</td>
</tr>
<tr class="even">
<td>interrupted</td>
<td>bool</td>
<td>False</td>
<td>is the syntenic cluster on multiple pseudomolecules?</td>
</tr>
<tr class="odd">
<td>palette</td>
<td>dict | None</td>
<td>{‘aliceblue’: ‘#F0F8FF’, ‘antiquewhite’: ‘#FAEBD7’, ‘aqua’: ‘#00FFFF’, ‘aquamarine’: ‘#7FFFD4’, ‘azure’: ‘#F0FFFF’, ‘beige’: ‘#F5F5DC’, ‘bisque’: ‘#FFE4C4’, ‘black’: ‘#000000’, ‘blanchedalmond’: ‘#FFEBCD’, ‘blue’: ‘#0000FF’, ‘blueviolet’: ‘#8A2BE2’, ‘brown’: ‘#A52A2A’, ‘burlywood’: ‘#DEB887’, ‘cadetblue’: ‘#5F9EA0’, ‘chartreuse’: ‘#7FFF00’, ‘chocolate’: ‘#D2691E’, ‘coral’: ‘#FF7F50’, ‘cornflowerblue’: ‘#6495ED’, ‘cornsilk’: ‘#FFF8DC’, ‘crimson’: ‘#DC143C’, ‘cyan’: ‘#00FFFF’, ‘darkblue’: ‘#00008B’, ‘darkcyan’: ‘#008B8B’, ‘darkgoldenrod’: ‘#B8860B’, ‘darkgray’: ‘#A9A9A9’, ‘darkgreen’: ‘#006400’, ‘darkgrey’: ‘#A9A9A9’, ‘darkkhaki’: ‘#BDB76B’, ‘darkmagenta’: ‘#8B008B’, ‘darkolivegreen’: ‘#556B2F’, ‘darkorange’: ‘#FF8C00’, ‘darkorchid’: ‘#9932CC’, ‘darkred’: ‘#8B0000’, ‘darksalmon’: ‘#E9967A’, ‘darkseagreen’: ‘#8FBC8F’, ‘darkslateblue’: ‘#483D8B’, ‘darkslategray’: ‘#2F4F4F’, ‘darkslategrey’: ‘#2F4F4F’, ‘darkturquoise’: ‘#00CED1’, ‘darkviolet’: ‘#9400D3’, ‘deeppink’: ‘#FF1493’, ‘deepskyblue’: ‘#00BFFF’, ‘dimgray’: ‘#696969’, ‘dimgrey’: ‘#696969’, ‘dodgerblue’: ‘#1E90FF’, ‘firebrick’: ‘#B22222’, ‘floralwhite’: ‘#FFFAF0’, ‘forestgreen’: ‘#228B22’, ‘fuchsia’: ‘#FF00FF’, ‘gainsboro’: ‘#DCDCDC’, ‘ghostwhite’: ‘#F8F8FF’, ‘gold’: ‘#FFD700’, ‘goldenrod’: ‘#DAA520’, ‘gray’: ‘#808080’, ‘green’: ‘#008000’, ‘greenyellow’: ‘#ADFF2F’, ‘grey’: ‘#808080’, ‘honeydew’: ‘#F0FFF0’, ‘hotpink’: ‘#FF69B4’, ‘indianred’: ‘#CD5C5C’, ‘indigo’: ‘#4B0082’, ‘ivory’: ‘#FFFFF0’, ‘khaki’: ‘#F0E68C’, ‘lavender’: ‘#E6E6FA’, ‘lavenderblush’: ‘#FFF0F5’, ‘lawngreen’: ‘#7CFC00’, ‘lemonchiffon’: ‘#FFFACD’, ‘lightblue’: ‘#ADD8E6’, ‘lightcoral’: ‘#F08080’, ‘lightcyan’: ‘#E0FFFF’, ‘lightgoldenrodyellow’: ‘#FAFAD2’, ‘lightgray’: ‘#D3D3D3’, ‘lightgreen’: ‘#90EE90’, ‘lightgrey’: ‘#D3D3D3’, ‘lightpink’: ‘#FFB6C1’, ‘lightsalmon’: ‘#FFA07A’, ‘lightseagreen’: ‘#20B2AA’, ‘lightskyblue’: ‘#87CEFA’, ‘lightslategray’: ‘#778899’, ‘lightslategrey’: ‘#778899’, ‘lightsteelblue’: ‘#B0C4DE’, ‘lightyellow’: ‘#FFFFE0’, ‘lime’: ‘#00FF00’, ‘limegreen’: ‘#32CD32’, ‘linen’: ‘#FAF0E6’, ‘magenta’: ‘#FF00FF’, ‘maroon’: ‘#800000’, ‘mediumaquamarine’: ‘#66CDAA’, ‘mediumblue’: ‘#0000CD’, ‘mediumorchid’: ‘#BA55D3’, ‘mediumpurple’: ‘#9370DB’, ‘mediumseagreen’: ‘#3CB371’, ‘mediumslateblue’: ‘#7B68EE’, ‘mediumspringgreen’: ‘#00FA9A’, ‘mediumturquoise’: ‘#48D1CC’, ‘mediumvioletred’: ‘#C71585’, ‘midnightblue’: ‘#191970’, ‘mintcream’: ‘#F5FFFA’, ‘mistyrose’: ‘#FFE4E1’, ‘moccasin’: ‘#FFE4B5’, ‘navajowhite’: ‘#FFDEAD’, ‘navy’: ‘#000080’, ‘oldlace’: ‘#FDF5E6’, ‘olive’: ‘#808000’, ‘olivedrab’: ‘#6B8E23’, ‘orange’: ‘#FFA500’, ‘orangered’: ‘#FF4500’, ‘orchid’: ‘#DA70D6’, ‘palegoldenrod’: ‘#EEE8AA’, ‘palegreen’: ‘#98FB98’, ‘paleturquoise’: ‘#AFEEEE’, ‘palevioletred’: ‘#DB7093’, ‘papayawhip’: ‘#FFEFD5’, ‘peachpuff’: ‘#FFDAB9’, ‘peru’: ‘#CD853F’, ‘pink’: ‘#FFC0CB’, ‘plum’: ‘#DDA0DD’, ‘powderblue’: ‘#B0E0E6’, ‘purple’: ‘#800080’, ‘rebeccapurple’: ‘#663399’, ‘red’: ‘#FF0000’, ‘rosybrown’: ‘#BC8F8F’, ‘royalblue’: ‘#4169E1’, ‘saddlebrown’: ‘#8B4513’, ‘salmon’: ‘#FA8072’, ‘sandybrown’: ‘#F4A460’, ‘seagreen’: ‘#2E8B57’, ‘seashell’: ‘#FFF5EE’, ‘sienna’: ‘#A0522D’, ‘silver’: ‘#C0C0C0’, ‘skyblue’: ‘#87CEEB’, ‘slateblue’: ‘#6A5ACD’, ‘slategray’: ‘#708090’, ‘slategrey’: ‘#708090’, ‘snow’: ‘#FFFAFA’, ‘springgreen’: ‘#00FF7F’, ‘steelblue’: ‘#4682B4’, ‘tan’: ‘#D2B48C’, ‘teal’: ‘#008080’, ‘thistle’: ‘#D8BFD8’, ‘tomato’: ‘#FF6347’, ‘turquoise’: ‘#40E0D0’, ‘violet’: ‘#EE82EE’, ‘wheat’: ‘#F5DEB3’, ‘white’: ‘#FFFFFF’, ‘whitesmoke’: ‘#F5F5F5’, ‘yellow’: ‘#FFFF00’, ‘yellowgreen’: ‘#9ACD32’}</td>
<td></td>
</tr>
<tr class="even">
<td>fontsize</td>
<td>int</td>
<td>30</td>
<td>this fontsize works well for default sizes, but is probably too big for smaller plots</td>
</tr>
<tr class="odd">
<td>chromosome_color</td>
<td>str</td>
<td>black</td>
<td>the color of the line representing the molecule the genes are plotted on.</td>
</tr>
<tr class="even">
<td>backup_gene_color</td>
<td>str</td>
<td>darkgray</td>
<td>If the GFF has no column titled “color”, the gene arrows will be filled using this color.</td>
</tr>
<tr class="odd">
<td>gene_name_offset</td>
<td>float</td>
<td>0.02</td>
<td>Default offset for plotting gene names if the GFF has no column titled “offset”.</td>
</tr>
<tr class="even">
<td>chromosome_width</td>
<td>float</td>
<td>3</td>
<td>Linewidth parameter for the line representing the chromosome; also, patch linewidth of the arrows representing genes. Please refer to the documentation of <code>matplotlib.patches.FancyArrowPath</code> for more details.</td>
</tr>
<tr class="odd">
<td>head_length</td>
<td>float</td>
<td>200</td>
<td>Head length for the arrows representing genes. Please refer to the documentation of <code>matplotlib.patches.FancyArrowPath</code> for more details.</td>
</tr>
<tr class="even">
<td>head_width</td>
<td>float</td>
<td>1.7</td>
<td>Head width for the arrows representing genes. Please refer to the documentation of <code>matplotlib.patches.FancyArrowPath</code> for more details.</td>
</tr>
<tr class="odd">
<td>gene_width</td>
<td>float</td>
<td>1</td>
<td>Width of the arrow patches representing genes. Please refer to the documentation of <code>matplotlib.patches.FancyArrowPath</code> for more details.</td>
</tr>
<tr class="even">
<td>ax</td>
<td>Axes</td>
<td>None</td>
<td>a pre-existing matplotlib axes object, if you wish to include this plot in a composite.</td>
</tr>
<tr class="odd">
<td>figsize</td>
<td>(&lt;class ‘float’&gt;, &lt;class ‘float’&gt;)</td>
<td>None</td>
<td></td>
</tr>
<tr class="even">
<td>save</td>
<td>str | pathlib.Path</td>
<td>schematic.svg</td>
<td>path to save an SVG image.</td>
</tr>
<tr class="odd">
<td>return_fig</td>
<td>bool</td>
<td>False</td>
<td>whether to return an axes object. Useful for further manipulation of the plot.</td>
</tr>
<tr class="even">
<td><strong>Returns</strong></td>
<td><strong>matplotlib.figure.Figure | matplotlib.axes._axes.Axes | None</strong></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="cell-14" class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"color"</span>] <span class="op">=</span> [</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"red"</span>,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"orange"</span>,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gold"</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"lightgreen"</span>,</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"forestgreen"</span>,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"royalblue"</span>,</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"darkblue"</span>,</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"darkmagenta"</span>,</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"magenta"</span>,</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"offset"</span>] <span class="op">=</span> <span class="fl">0.02</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"strand"</span>] <span class="op">=</span> <span class="st">"-"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Sometimes (many times) we are not interested in plotting syntenic genes with their real distances; rather, we are interested in the order and orientation of the genes. This representation can be achieved with the <a href="https://galicae.github.io/geneorder/core.html#plot_synteny_schematic"><code>plot_synteny_schematic</code></a> function, that can process the same sort of input as <a href="https://galicae.github.io/geneorder/core.html#plot_synteny"><code>plot_synteny</code></a>.</p>
<div id="cell-16" class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(hox)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-11-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The strand is reflected on the orientation of the arrows - genes on the plus strand point right.</p>
<div id="cell-18" class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"strand"</span>] <span class="op">=</span> <span class="st">"+"</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(hox)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-12-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The <code>geneorder</code> package uses <code>matplotlib.patches.FancyArrow</code> for the schematic plots, and hence the shape of the arrowhead can be precisely controlled:</p>
<div id="cell-20" class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(hox, head_length<span class="op">=</span><span class="dv">100</span>, head_width<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(hox, head_length<span class="op">=</span><span class="dv">300</span>, head_width<span class="op">=</span><span class="fl">1.0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-13-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-13-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>An interesting edge case is setting the head width and length both to <code>0</code>, in order to plot rectangles. This may be useful if the orientation of the features is not important:</p>
<div id="cell-22" class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(hox, head_length<span class="op">=</span><span class="dv">0</span>, head_width<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-14-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Sometimes we will not have information about the orientation of single genes, for instance when there is no gene model predicted. In those cases, the <code>strand</code> can be set to <code>.</code>, something the GFF standard often uses to denote missing values. The width of the block will still be equivalent to the arrows, but no arrowhead will be plotted.</p>
<div id="cell-24" class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"strand"</span>] <span class="op">=</span> [<span class="st">"+"</span>, <span class="st">"+"</span>, <span class="st">"+"</span>, <span class="st">"+"</span>, <span class="st">"."</span>, <span class="st">"."</span>, <span class="st">"-"</span>, <span class="st">"-"</span>, <span class="st">"+"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-25" class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(hox)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-16-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Similarly, we may want to highlight one or multiple of the genes, for instance because they were annotated via a different route. <code>geneorder</code> accepts an <code>edge_color</code> parameter in the input “GFF”, and will plot black edges if not otherwise instructed.</p>
<p>Let’s say we want to highlight all genes that were based on PacBio transcripts:</p>
<div id="cell-27" class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"edge_color"</span>] <span class="op">=</span> hox[<span class="st">"source"</span>].replace({<span class="st">"PacBio"</span>: <span class="st">"cyan"</span>, <span class="st">"AUGUSTUS"</span>: <span class="st">"black"</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-28" class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(hox)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-18-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The keen reader will have noticed that there is no gap between Hox8 and Hox10 - even though Hox9 is missing. We can edit the GFF dataframe to amend this:</p>
<div id="cell-30" class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"strand"</span>] <span class="op">=</span> <span class="st">"+"</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>hox[<span class="st">"edge_color"</span>] <span class="op">=</span> <span class="st">"black"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-31" class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>hox <span class="op">=</span> util.insert_gap(</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    hox,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"g9724"</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"g9725"</span>,</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"gene_id"</span>,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>    no_gaps<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>    purge_columns<span class="op">=</span>[<span class="st">"gene_name"</span>, <span class="st">"color"</span>, <span class="st">"offset"</span>, <span class="st">"attributes"</span>],</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-32" class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>hox</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">seqid</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">strand</th>
<th data-quarto-table-cell-role="th">phase</th>
<th data-quarto-table-cell-role="th">attributes</th>
<th data-quarto-table-cell-role="th">gene_id</th>
<th data-quarto-table-cell-role="th">gene_name</th>
<th data-quarto-table-cell-role="th">color</th>
<th data-quarto-table-cell-role="th">offset</th>
<th data-quarto-table-cell-role="th">edge_color</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>pseudochrom_56</td>
<td>PacBio</td>
<td>gene</td>
<td>1927066</td>
<td>1936157</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=PB.8615;function=Homeobox domain;gene=Hox1-...</td>
<td>PB.8615</td>
<td>Hox1-A</td>
<td>red</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>1998922</td>
<td>2024148</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=g9718;function=sequence-specific DNA bindin...</td>
<td>g9718</td>
<td>Hox2-A</td>
<td>orange</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>pseudochrom_56</td>
<td>PacBio</td>
<td>gene</td>
<td>2058396</td>
<td>2065953</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=PB.8616;function=homeobox protein;gene=Hox3...</td>
<td>PB.8616</td>
<td>Hox3-A</td>
<td>gold</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2195412</td>
<td>2206712</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=g9720;function=sequence-specific DNA bindin...</td>
<td>g9720</td>
<td>Hox4-A</td>
<td>lightgreen</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2351936</td>
<td>2354374</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=g9721;function=sequence-specific DNA bindin...</td>
<td>g9721</td>
<td>Hox5-A</td>
<td>forestgreen</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>pseudochrom_56</td>
<td>PacBio</td>
<td>gene</td>
<td>2373415</td>
<td>2375678</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=PB.8617;function=sequence-specific DNA bind...</td>
<td>PB.8617</td>
<td>Hox6-A</td>
<td>royalblue</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2565196</td>
<td>2594468</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=g9723;function=sequence-specific DNA bindin...</td>
<td>g9723</td>
<td>Hox7-A</td>
<td>darkblue</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2916314</td>
<td>2926445</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=g9724;function=sequence-specific DNA bindin...</td>
<td>g9724</td>
<td>Hox8-A</td>
<td>darkmagenta</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2926446</td>
<td>2926447</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td></td>
<td>gap_g9724-0</td>
<td></td>
<td></td>
<td></td>
<td>black</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2986021</td>
<td>2996225</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=g9725;function=sequence-specific DNA bindin...</td>
<td>g9725</td>
<td>Hox10-A</td>
<td>magenta</td>
<td>0.02</td>
<td>black</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-33" class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(hox)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-22-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>This approach becomes even more powerful when we are plotting the same cluster in different instances, for example when there is a duplication of the cluster:</p>
<div id="cell-35" class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>on_chrom12 <span class="op">=</span> gff[<span class="st">"seqid"</span>] <span class="op">==</span> <span class="st">"pseudochrom_12"</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>is_gene <span class="op">=</span> gff[<span class="st">"type"</span>] <span class="op">==</span> <span class="st">"gene"</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>hoxb <span class="op">=</span> gff[on_chrom12 <span class="op">&amp;</span> is_gene].copy().reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>hoxb[<span class="st">"color"</span>] <span class="op">=</span> [<span class="st">"red"</span>, <span class="st">"orange"</span>, <span class="st">"darkblue"</span>, <span class="st">"magenta"</span>]</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>hoxb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">seqid</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">strand</th>
<th data-quarto-table-cell-role="th">phase</th>
<th data-quarto-table-cell-role="th">attributes</th>
<th data-quarto-table-cell-role="th">gene_id</th>
<th data-quarto-table-cell-role="th">gene_name</th>
<th data-quarto-table-cell-role="th">color</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>pseudochrom_12</td>
<td>PacBio</td>
<td>gene</td>
<td>1927066</td>
<td>1936157</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=PB.9615;function=Homeobox domain;gene=Hox1-...</td>
<td>PB.9615</td>
<td>Hox1-B</td>
<td>red</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>1998922</td>
<td>2024148</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7718;function=sequence-specific DNA bindin...</td>
<td>g7718</td>
<td>Hox2-B</td>
<td>orange</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2565196</td>
<td>2594468</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7723;function=sequence-specific DNA bindin...</td>
<td>g7723</td>
<td>Hox7-B</td>
<td>darkblue</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2986021</td>
<td>2996225</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7725;function=sequence-specific DNA bindin...</td>
<td>g7725</td>
<td>Hox10-B</td>
<td>magenta</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>This cluster is missing Hox genes 3-6 and 8-9; we should add the corresponding number of gaps in order for the plots to line up accordingly:</p>
<div id="cell-37" class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>hoxb <span class="op">=</span> util.insert_gap(</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    hoxb, <span class="st">"g7718"</span>, <span class="st">"g7723"</span>, <span class="st">"gene_id"</span>, no_gaps<span class="op">=</span><span class="dv">4</span>, purge_columns<span class="op">=</span>[<span class="st">"gene_name"</span>, <span class="st">"color"</span>]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>hoxb <span class="op">=</span> util.insert_gap(</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    hoxb, <span class="st">"g7723"</span>, <span class="st">"g7725"</span>, <span class="st">"gene_id"</span>, no_gaps<span class="op">=</span><span class="dv">2</span>, purge_columns<span class="op">=</span>[<span class="st">"gene_name"</span>, <span class="st">"color"</span>]</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>hoxb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">seqid</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">strand</th>
<th data-quarto-table-cell-role="th">phase</th>
<th data-quarto-table-cell-role="th">attributes</th>
<th data-quarto-table-cell-role="th">gene_id</th>
<th data-quarto-table-cell-role="th">gene_name</th>
<th data-quarto-table-cell-role="th">color</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>pseudochrom_12</td>
<td>PacBio</td>
<td>gene</td>
<td>1927066</td>
<td>1936157</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=PB.9615;function=Homeobox domain;gene=Hox1-...</td>
<td>PB.9615</td>
<td>Hox1-B</td>
<td>red</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>1998922</td>
<td>2024148</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7718;function=sequence-specific DNA bindin...</td>
<td>g7718</td>
<td>Hox2-B</td>
<td>orange</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2024149</td>
<td>2024150</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7718;function=sequence-specific DNA bindin...</td>
<td>gap_g7718-0</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2024151</td>
<td>2024152</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7718;function=sequence-specific DNA bindin...</td>
<td>gap_g7718-1</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2024153</td>
<td>2024154</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7718;function=sequence-specific DNA bindin...</td>
<td>gap_g7718-2</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2024155</td>
<td>2024156</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7718;function=sequence-specific DNA bindin...</td>
<td>gap_g7718-3</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2565196</td>
<td>2594468</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7723;function=sequence-specific DNA bindin...</td>
<td>g7723</td>
<td>Hox7-B</td>
<td>darkblue</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2594469</td>
<td>2594470</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7723;function=sequence-specific DNA bindin...</td>
<td>gap_g7723-0</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2594471</td>
<td>2594472</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7723;function=sequence-specific DNA bindin...</td>
<td>gap_g7723-1</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>pseudochrom_12</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2986021</td>
<td>2996225</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g7725;function=sequence-specific DNA bindin...</td>
<td>g7725</td>
<td>Hox10-B</td>
<td>magenta</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-38" class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(hox)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(hoxb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-25-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-25-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Sometimes we want to plot genes that belong to the same cluster but because of assembly fragmentation they are not on the same pseudomolecule.</p>
<div id="cell-40" class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>on_scaff44 <span class="op">=</span> gff[<span class="st">"seqid"</span>] <span class="op">==</span> <span class="st">"scaffold_44"</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>is_gene <span class="op">=</span> gff[<span class="st">"type"</span>] <span class="op">==</span> <span class="st">"gene"</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>hoxc <span class="op">=</span> gff[on_scaff44 <span class="op">&amp;</span> is_gene].copy().reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>hoxc[<span class="st">"color"</span>] <span class="op">=</span> [<span class="st">"dodgerblue"</span>, <span class="st">"hotpink"</span>]</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>hoxc[<span class="st">"edge_color"</span>] <span class="op">=</span> <span class="st">"black"</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>interrupted <span class="op">=</span> pd.concat((hox.loc[<span class="dv">6</span>:], hoxc)).reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>interrupted[<span class="st">"offset"</span>] <span class="op">=</span> <span class="fl">0.02</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>interrupted <span class="op">=</span> util.insert_break(interrupted, locus1<span class="op">=</span><span class="st">"g9725"</span>, locus2<span class="op">=</span><span class="st">"PB.1762"</span>)</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>interrupted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">seqid</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">type</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">score</th>
<th data-quarto-table-cell-role="th">strand</th>
<th data-quarto-table-cell-role="th">phase</th>
<th data-quarto-table-cell-role="th">attributes</th>
<th data-quarto-table-cell-role="th">gene_id</th>
<th data-quarto-table-cell-role="th">gene_name</th>
<th data-quarto-table-cell-role="th">color</th>
<th data-quarto-table-cell-role="th">offset</th>
<th data-quarto-table-cell-role="th">edge_color</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2565196</td>
<td>2594468</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=g9723;function=sequence-specific DNA bindin...</td>
<td>g9723</td>
<td>Hox7-A</td>
<td>darkblue</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2916314</td>
<td>2926445</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=g9724;function=sequence-specific DNA bindin...</td>
<td>g9724</td>
<td>Hox8-A</td>
<td>darkmagenta</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2926446</td>
<td>2926447</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td></td>
<td>gap_g9724-0</td>
<td></td>
<td></td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>pseudochrom_56</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>2986021</td>
<td>2996225</td>
<td>.</td>
<td>+</td>
<td>.</td>
<td>ID=g9725;function=sequence-specific DNA bindin...</td>
<td>g9725</td>
<td>Hox10-A</td>
<td>magenta</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td></td>
<td>break</td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>scaffold_44</td>
<td>PacBio</td>
<td>gene</td>
<td>1927066</td>
<td>1936157</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=PB.1762;function=Homeobox domain;gene=Hox11...</td>
<td>PB.1762</td>
<td>Hox11</td>
<td>dodgerblue</td>
<td>0.02</td>
<td>black</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>scaffold_44</td>
<td>AUGUSTUS</td>
<td>gene</td>
<td>1998922</td>
<td>2024148</td>
<td>.</td>
<td>-</td>
<td>.</td>
<td>ID=g13061;function=sequence-specific DNA bindi...</td>
<td>g13061</td>
<td>Hox12</td>
<td>hotpink</td>
<td>0.02</td>
<td>black</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-41" class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>plot_synteny_schematic(interrupted, interrupted<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="00_core_files/figure-html/cell-27-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/galicae\.github\.io\/geneorder");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/galicae/geneorder/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>