# goppy


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

This package was born out of my continuous efforts to write streamlined
code to plot Hox gene clusters. It seems like this is a task I will be
performing many times, and there don’t seem to be too many tools out
there, so I wrote my own. This is my `matplotlib`-based answer, and I
hope that it is useful to you, too.

## Input

Generally, the package expects you to have, at least, knowledge of the
gene IDs that you want to visualize, and knowledge of their coordinates
in the genome. This can be manually encoded or come in table form -
crucially, it can be read from a GFF3 file.

## Usage

### Installation

Install latest from the GitHub
[repository](https://github.com/galicae/goppy):

``` sh
$ pip install git+https://github.com/galicae/goppy.git
```

or from [conda](https://anaconda.org/galicae/goppy)

``` sh
$ conda install -c galicae goppy
```

or from [pypi](https://pypi.org/project/goppy/)

``` sh
$ pip install goppy
```

### Documentation

Documentation can be found hosted on this GitHub
[repository](https://github.com/galicae/goppy)’s
[pages](https://galicae.github.io/goppy/). Additionally you can find
package manager specific guidelines on
[conda](https://anaconda.org/galicae/goppy) and
[pypi](https://pypi.org/project/goppy/) respectively.

## How to use

The most basic `goppy` use case is when you have a list of genes that
are co-linear and would like to plot them in their chromosomal context.

``` python
1 + 1
```

    2
