{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# util\n",
    "\n",
    "> I/O functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | default_exp util"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "\n",
    "from fastcore.docments import docments\n",
    "from fastcore.test import *\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "import os\n",
    "from collections.abc import Sequence\n",
    "from pathlib import Path\n",
    "from typing import Literal\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The information about the location of genes in a genome can come from anywhere, but it is typically stored in a GFF3 file. While this is far from a standardized format, there exist some _guidelines_ about how a (syntactically) valid GFF3 file should look like; they can be found [here](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md). Such a typical, 9-column, tab-separated file is what I have mostly been working with. If your GFF3 looks different you will have to figure out how to read it in yourself - or [open an issue](https://github.com/galicae/geneorder/issues/new) and I can try to help you."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def read_gff(\n",
    "    loc: str | Path,  # input filepath\n",
    "    gff_columns: list = [\n",
    "        \"seqid\",\n",
    "        \"source\",\n",
    "        \"type\",\n",
    "        \"start\",\n",
    "        \"end\",\n",
    "        \"score\",\n",
    "        \"strand\",\n",
    "        \"phase\",\n",
    "        \"attributes\",\n",
    "    ],  # column names for the GFF3 file\n",
    "    skiprows: int = 1,  # how many rows to skip in the beginning\n",
    "    header: (\n",
    "        int | Sequence[int] | None | Literal[\"infer\"]\n",
    "    ) = None,  # whether to expect a header or not\n",
    "    sep: str = \"\\t\",  # separator for the table\n",
    "    **kwargs,  # various pd.read_csv() arguments\n",
    ") -> pd.DataFrame:  # the GFF3 file in DataFrame form\n",
    "    \"A function to read a GFF3 file. Expects 9 tab-separated fields, and will try to name the columns according to https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md\"\n",
    "    gff = pd.read_csv(loc, sep=sep, header=header, skiprows=skiprows, **kwargs)\n",
    "    gff.columns = gff_columns\n",
    "    return gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For instance, let's try and read an example. This is a fake GFF3 file based on the _Pycnogonum litorale_ Hox gene cluster; for the real one, please refer to our [work on the sea spider](https://doi.org/10.1101/2024.11.20.624475)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "if \"EXAMPLE_DATA_PATH\" not in os.environ.keys():\n",
    "    os.environ[\"EXAMPLE_DATA_PATH\"] = (\n",
    "        \"/Users/npapadop/Documents/repos/geneorder/example_data/\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>1927066</td>\n",
       "      <td>1936157</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8615;function=Homeobox domain;gene=Hox1-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1927066</td>\n",
       "      <td>1936157</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8615.1;Parent=PB.8615;function=Homeobox ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1927066</td>\n",
       "      <td>1928028</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8615.1.exon1;Parent=PB.8615.1;function=H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>exon</td>\n",
       "      <td>1935229</td>\n",
       "      <td>1936157</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8615.1.exon2;Parent=PB.8615.1;function=H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1927066</td>\n",
       "      <td>1928028</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>ID=PB.8615.1.CDS1;Parent=PB.8615.1;function=Ho...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>pseudochrom_12</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>2986021</td>\n",
       "      <td>2996225</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g7725.t1;Parent=g7725;function=sequence-spe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>pseudochrom_12</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>exon</td>\n",
       "      <td>2986021</td>\n",
       "      <td>2986325</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g7725.t1.exon1;Parent=g7725.t1;function=seq...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>pseudochrom_12</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>exon</td>\n",
       "      <td>2995445</td>\n",
       "      <td>2996225</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g7725.t1.exon2;Parent=g7725.t1;function=seq...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>pseudochrom_12</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2986021</td>\n",
       "      <td>2986325</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ID=g7725.t1.CDS1;Parent=g7725.t1;function=sequ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>pseudochrom_12</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2995445</td>\n",
       "      <td>2996225</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=g7725.t1.CDS2;Parent=g7725.t1;function=sequ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>124 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              seqid    source  type    start      end score strand phase  \\\n",
       "0    pseudochrom_56    PacBio  gene  1927066  1936157     .      -     .   \n",
       "1    pseudochrom_56    PacBio  mRNA  1927066  1936157     .      -     .   \n",
       "2    pseudochrom_56    PacBio  exon  1927066  1928028     .      -     .   \n",
       "3    pseudochrom_56    PacBio  exon  1935229  1936157     .      -     .   \n",
       "4    pseudochrom_56    PacBio   CDS  1927066  1928028     .      -     1   \n",
       "..              ...       ...   ...      ...      ...   ...    ...   ...   \n",
       "119  pseudochrom_12  AUGUSTUS  mRNA  2986021  2996225     1      -     .   \n",
       "120  pseudochrom_12  AUGUSTUS  exon  2986021  2986325     .      -     .   \n",
       "121  pseudochrom_12  AUGUSTUS  exon  2995445  2996225     .      -     .   \n",
       "122  pseudochrom_12  AUGUSTUS   CDS  2986021  2986325     1      -     2   \n",
       "123  pseudochrom_12  AUGUSTUS   CDS  2995445  2996225     1      -     0   \n",
       "\n",
       "                                            attributes  \n",
       "0    ID=PB.8615;function=Homeobox domain;gene=Hox1-...  \n",
       "1    ID=PB.8615.1;Parent=PB.8615;function=Homeobox ...  \n",
       "2    ID=PB.8615.1.exon1;Parent=PB.8615.1;function=H...  \n",
       "3    ID=PB.8615.1.exon2;Parent=PB.8615.1;function=H...  \n",
       "4    ID=PB.8615.1.CDS1;Parent=PB.8615.1;function=Ho...  \n",
       "..                                                 ...  \n",
       "119  ID=g7725.t1;Parent=g7725;function=sequence-spe...  \n",
       "120  ID=g7725.t1.exon1;Parent=g7725.t1;function=seq...  \n",
       "121  ID=g7725.t1.exon2;Parent=g7725.t1;function=seq...  \n",
       "122  ID=g7725.t1.CDS1;Parent=g7725.t1;function=sequ...  \n",
       "123  ID=g7725.t1.CDS2;Parent=g7725.t1;function=sequ...  \n",
       "\n",
       "[124 rows x 9 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff = read_gff(os.environ[\"EXAMPLE_DATA_PATH\"] + \"plit.gff3\")\n",
    "gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To begin with, we would like to visualize the 'real' Hox genes. We are also only interested in plotting entire genes (no exon structure), so we can filter the GFF rows based on that. We should also extract the gene IDs from the table to help us filter. Furthermore, we should be extracting gene names, in case we want to plot them too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def gff_attribute_selector(\n",
    "    line,  # the attributes field of a GFF3 line\n",
    "    sep: str = \";\",  # the field separator. Should be a semicolon for a GFF3 file.\n",
    "    select=\"ID\",  # the field ID. Should be one that is included in the GFF3 file. Refer to https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md for possibilities, or choose a manually defined tag that you know is present in the file.\n",
    ") -> str:  # the value of field `select`\n",
    "    \"A function that extracts the value of a specified field from the attributes of a GFF3 line. Should be used on the `attributes` field of the corresponding pandas DataFrame.\"\n",
    "    for field in line.split(sep):\n",
    "        if field.startswith(select):\n",
    "            value = field.split(\"=\")[1]\n",
    "            return value\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff[\"gene_id\"] = gff[\"attributes\"].apply(\n",
    "    lambda x: gff_attribute_selector(x, select=\"ID\")\n",
    ")\n",
    "gff[\"gene_name\"] = gff[\"attributes\"].apply(\n",
    "    lambda x: gff_attribute_selector(x, select=\"gene\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>1927066</td>\n",
       "      <td>1936157</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8615;function=Homeobox domain;gene=Hox1-...</td>\n",
       "      <td>PB.8615</td>\n",
       "      <td>Hox1-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>1998922</td>\n",
       "      <td>2024148</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g9718;function=sequence-specific DNA bindin...</td>\n",
       "      <td>g9718</td>\n",
       "      <td>Hox2-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>2058396</td>\n",
       "      <td>2065953</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8616;function=homeobox protein;gene=Hox3...</td>\n",
       "      <td>PB.8616</td>\n",
       "      <td>Hox3-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>2195412</td>\n",
       "      <td>2206712</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g9720;function=sequence-specific DNA bindin...</td>\n",
       "      <td>g9720</td>\n",
       "      <td>Hox4-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>2351936</td>\n",
       "      <td>2354374</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g9721;function=sequence-specific DNA bindin...</td>\n",
       "      <td>g9721</td>\n",
       "      <td>Hox5-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>2373415</td>\n",
       "      <td>2375678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8617;function=sequence-specific DNA bind...</td>\n",
       "      <td>PB.8617</td>\n",
       "      <td>Hox6-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>2565196</td>\n",
       "      <td>2594468</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g9723;function=sequence-specific DNA bindin...</td>\n",
       "      <td>g9723</td>\n",
       "      <td>Hox7-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>2916314</td>\n",
       "      <td>2926445</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g9724;function=sequence-specific DNA bindin...</td>\n",
       "      <td>g9724</td>\n",
       "      <td>Hox8-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>2986021</td>\n",
       "      <td>2996225</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g9725;function=sequence-specific DNA bindin...</td>\n",
       "      <td>g9725</td>\n",
       "      <td>Hox10-A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqid    source  type    start      end score strand phase  \\\n",
       "0   pseudochrom_56    PacBio  gene  1927066  1936157     .      -     .   \n",
       "6   pseudochrom_56  AUGUSTUS  gene  1998922  2024148     .      -     .   \n",
       "15  pseudochrom_56    PacBio  gene  2058396  2065953     .      -     .   \n",
       "56  pseudochrom_56  AUGUSTUS  gene  2195412  2206712     .      -     .   \n",
       "62  pseudochrom_56  AUGUSTUS  gene  2351936  2354374     .      -     .   \n",
       "68  pseudochrom_56    PacBio  gene  2373415  2375678     .      -     .   \n",
       "79  pseudochrom_56  AUGUSTUS  gene  2565196  2594468     .      -     .   \n",
       "85  pseudochrom_56  AUGUSTUS  gene  2916314  2926445     .      -     .   \n",
       "91  pseudochrom_56  AUGUSTUS  gene  2986021  2996225     .      -     .   \n",
       "\n",
       "                                           attributes  gene_id gene_name  \n",
       "0   ID=PB.8615;function=Homeobox domain;gene=Hox1-...  PB.8615    Hox1-A  \n",
       "6   ID=g9718;function=sequence-specific DNA bindin...    g9718    Hox2-A  \n",
       "15  ID=PB.8616;function=homeobox protein;gene=Hox3...  PB.8616    Hox3-A  \n",
       "56  ID=g9720;function=sequence-specific DNA bindin...    g9720    Hox4-A  \n",
       "62  ID=g9721;function=sequence-specific DNA bindin...    g9721    Hox5-A  \n",
       "68  ID=PB.8617;function=sequence-specific DNA bind...  PB.8617    Hox6-A  \n",
       "79  ID=g9723;function=sequence-specific DNA bindin...    g9723    Hox7-A  \n",
       "85  ID=g9724;function=sequence-specific DNA bindin...    g9724    Hox8-A  \n",
       "91  ID=g9725;function=sequence-specific DNA bindin...    g9725   Hox10-A  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hox_genes = [\n",
    "    \"PB.8615\",\n",
    "    \"g9718\",\n",
    "    \"PB.8616\",\n",
    "    \"g9720\",\n",
    "    \"g9721\",\n",
    "    \"PB.8617\",\n",
    "    \"g9723\",\n",
    "    \"g9724\",\n",
    "    \"g9725\",\n",
    "]\n",
    "is_hox = gff[\"gene_id\"].isin(hox_genes)\n",
    "is_gene = gff[\"type\"] == \"gene\"\n",
    "\n",
    "slim = gff[is_gene & is_hox]\n",
    "slim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This table basically already contains all the information we need: \n",
    "* the name of the chromosome\n",
    "* the location of each gene\n",
    "* the strand of each gene\n",
    "* the directionality of each gene (given by relative start/end positions)\n",
    "* the name of each gene (hidden in the attributes)\n",
    "\n",
    "The only thing that's left is to extract this information in the way that is needed for the plotter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox1-A</td>\n",
       "      <td>PB.8615</td>\n",
       "      <td>1927066</td>\n",
       "      <td>1936157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox2-A</td>\n",
       "      <td>g9718</td>\n",
       "      <td>1998922</td>\n",
       "      <td>2024148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox3-A</td>\n",
       "      <td>PB.8616</td>\n",
       "      <td>2058396</td>\n",
       "      <td>2065953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox4-A</td>\n",
       "      <td>g9720</td>\n",
       "      <td>2195412</td>\n",
       "      <td>2206712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox5-A</td>\n",
       "      <td>g9721</td>\n",
       "      <td>2351936</td>\n",
       "      <td>2354374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox6-A</td>\n",
       "      <td>PB.8617</td>\n",
       "      <td>2373415</td>\n",
       "      <td>2375678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox7-A</td>\n",
       "      <td>g9723</td>\n",
       "      <td>2565196</td>\n",
       "      <td>2594468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox8-A</td>\n",
       "      <td>g9724</td>\n",
       "      <td>2916314</td>\n",
       "      <td>2926445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox10-A</td>\n",
       "      <td>g9725</td>\n",
       "      <td>2986021</td>\n",
       "      <td>2996225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqid gene_name  gene_id    start      end\n",
       "0  pseudochrom_56    Hox1-A  PB.8615  1927066  1936157\n",
       "1  pseudochrom_56    Hox2-A    g9718  1998922  2024148\n",
       "2  pseudochrom_56    Hox3-A  PB.8616  2058396  2065953\n",
       "3  pseudochrom_56    Hox4-A    g9720  2195412  2206712\n",
       "4  pseudochrom_56    Hox5-A    g9721  2351936  2354374\n",
       "5  pseudochrom_56    Hox6-A  PB.8617  2373415  2375678\n",
       "6  pseudochrom_56    Hox7-A    g9723  2565196  2594468\n",
       "7  pseudochrom_56    Hox8-A    g9724  2916314  2926445\n",
       "8  pseudochrom_56   Hox10-A    g9725  2986021  2996225"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hox = slim[[\"seqid\", \"gene_name\", \"gene_id\", \"start\", \"end\"]].reset_index(drop=True)\n",
    "hox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def decorate(\n",
    "    gff: pd.DataFrame,  # a GFF file in Pandas dataframe form\n",
    "    attributes: dict = {\n",
    "        \"gene_id\": \"ID\",\n",
    "        \"gene_name\": \"gene\",\n",
    "    },  # a dictionary of tags to extract from the attributes column. For each key/value pair, the value under the `key` tag will be saved in a new column named `value`.\n",
    "):\n",
    "    \"A function that\"\n",
    "    for attr_add, attr_filter in attributes.items():\n",
    "        gff[attr_add] = gff[\"attributes\"].apply(\n",
    "            lambda x: gff_attribute_selector(x, select=attr_filter)\n",
    "        )\n",
    "\n",
    "\n",
    "# | export\n",
    "\n",
    "\n",
    "def filter(\n",
    "    gff,\n",
    "    filter_by_type=True,\n",
    "    filter_type=\"gene\",\n",
    "    filter_by_field=True,\n",
    "    field=\"gene_id\",\n",
    "    field_values=None,\n",
    "):\n",
    "    if filter_type and filter_type not in gff[filter_type]:\n",
    "        raise ValueError(\n",
    "            f'The value {filter_type} is not found in the \"type\" column of the input table.'\n",
    "        )\n",
    "\n",
    "    if filter_by_field:\n",
    "        if field not in gff.columns:\n",
    "            raise ValueError(\n",
    "                f'The value {filter_type} is not found in the \"type\" column of the input table.'\n",
    "            )\n",
    "\n",
    "    keep = np.array([True] * gff.shape[0])\n",
    "    if genes is not None:\n",
    "        is_genes = gff[\"gene_id\"].isin(genes)\n",
    "        keep = keep & is_genes\n",
    "\n",
    "    if is_type:\n",
    "        is_gene = gff[\"type\"] == \"gene\"\n",
    "\n",
    "    slim = gff[is_gene & is_hox]\n",
    "    return slim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def syntenic_block_borders(\n",
    "    gff: pd.DataFrame,  # a GFF in Pandas dataframe form. Only includes the genes of the syntenic block in question.\n",
    "    flank_length: int = None,  # the amount of space to be granted on both sides of the syntenic region, in basepairs. If unspecified, it will be set to 5% of the syntenic block length.\n",
    "    start: str = \"start\",  # the GFF column with the start position of the gene (\"start\").\n",
    "    end: str = \"end\",  # the GFF column with the end position of the gene (\"end\").\n",
    ") -> (int, int):\n",
    "    \"A function to calculate the boundaries of a syntenic block. It automatically pads the boundaries by an additional 5% of total length on both ends.\"\n",
    "    possible = [gff[start].min(), gff[start].max(), gff[end].min(), gff[end].max()]\n",
    "    block_start = min(possible)\n",
    "    block_end = max(possible)\n",
    "    if flank_length is None:\n",
    "        flank_length = (block_end - block_start) // 100 * 5\n",
    "    if not flank_length.is_integer():\n",
    "        raise ValueError(\n",
    "            f'The parameter `flank_length` has to be an integer. You supplied the value \"{flank_length}\" which is {type(flank_length)}.'\n",
    "        )\n",
    "    return max(block_start - flank_length, 0), block_end + flank_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_fail(\n",
    "    syntenic_block_borders,\n",
    "    contains=\"The parameter `flank_length` has to be an integer\",\n",
    "    args=(gff,),\n",
    "    kwargs=dict(flank_length=5.2),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The process of reading in a GFF3 file can be expedited with the "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def read_aln(\n",
    "    m8: str,  # the path to the MMseqs2 alignment table file\n",
    "    id_sep: (\n",
    "        str | None\n",
    "    ) = None,  # (optional) a character that separates the species ID from the gene ID\n",
    "    **kwargs,  # various arguments to be passed to the pd.read_csv() function\n",
    ") -> pd.DataFrame:  # the tabulated form of the alignment results.\n",
    "    \"Reads\"\n",
    "    hits = pd.read_csv(m8, sep=\"\\t\", header=None)\n",
    "    m8_columns = [\n",
    "        \"query\",\n",
    "        \"target\",\n",
    "        \"seq_id\",\n",
    "        \"ali_len\",\n",
    "        \"no_mism\",\n",
    "        \"no_go\",\n",
    "        \"q_start\",\n",
    "        \"q_end\",\n",
    "        \"t_start\",\n",
    "        \"t_end\",\n",
    "        \"eval\",\n",
    "        \"bit\",\n",
    "    ]\n",
    "    hits.columns = m8_columns\n",
    "    # trim the query to just the ID\n",
    "    if id_sep is not None:\n",
    "        hits[\"query\"] = hits[\"query\"].str.split(id_sep).str[1].str.join(id_sep)\n",
    "    return hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "\n",
    "\n",
    "def estimate_plot_size(\n",
    "    gff,\n",
    "    width_factor: int = 3,\n",
    "    height: int = 2,\n",
    "):\n",
    "    x = len(gff) * width_factor\n",
    "    return (x, height)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert estimate_plot_size(gff[gff[\"type\"] == \"gene\"]) == (39, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | export\n",
    "def insert_gap(\n",
    "    gff: pd.DataFrame,\n",
    "    locus1,\n",
    "    locus2,\n",
    "    identifier,\n",
    "    purge_columns=None,\n",
    "    no_gaps=1,\n",
    ") -> pd.DataFrame:\n",
    "    \"This function inserts a number of dummy entries between two loci (lines) in the GFF DataFrame.\"\n",
    "    loc1_index = gff[gff[identifier] == locus1].index[0]\n",
    "    loc2_index = gff[gff[identifier] == locus2].index[0]\n",
    "    if loc2_index - loc1_index != 1:\n",
    "        raise ValueError(\n",
    "            f\"The two loci are not consecutive; their indices are {locus1}: {loc1_index} and {locus2}: {loc2_index}, respectively. Please refer to your input dataframe.\"\n",
    "        )\n",
    "\n",
    "    offset_step = 1 / (no_gaps + 1)\n",
    "\n",
    "    for i in range(no_gaps):\n",
    "        offset = (i + 1) * offset_step\n",
    "        tmp = gff.loc[loc1_index].copy()\n",
    "        tmp[\"start\"] = tmp[\"end\"] + 1 + (i * 2)\n",
    "        tmp[\"end\"] = tmp[\"start\"] + 1\n",
    "        tmp[identifier] = f\"gap_{locus1}-{i}\"\n",
    "        for column in purge_columns:\n",
    "            tmp[column] = \"\"\n",
    "        gff.loc[loc1_index + offset] = tmp\n",
    "\n",
    "    gff = gff.sort_index().reset_index(drop=True)\n",
    "    return gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_fail(\n",
    "    insert_gap,\n",
    "    contains=\"The two loci are not consecutive;\",\n",
    "    args=(hox, \"PB.8615\", \"g9720\", \"gene_id\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hox = insert_gap(\n",
    "    hox, \"PB.8615\", \"g9718\", \"gene_id\", no_gaps=4, purge_columns=[\"gene_name\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox1-A</td>\n",
       "      <td>PB.8615</td>\n",
       "      <td>1927066</td>\n",
       "      <td>1936157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td></td>\n",
       "      <td>gap_PB.8615-0</td>\n",
       "      <td>1936158</td>\n",
       "      <td>1936159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td></td>\n",
       "      <td>gap_PB.8615-1</td>\n",
       "      <td>1936160</td>\n",
       "      <td>1936161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td></td>\n",
       "      <td>gap_PB.8615-2</td>\n",
       "      <td>1936162</td>\n",
       "      <td>1936163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td></td>\n",
       "      <td>gap_PB.8615-3</td>\n",
       "      <td>1936164</td>\n",
       "      <td>1936165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox2-A</td>\n",
       "      <td>g9718</td>\n",
       "      <td>1998922</td>\n",
       "      <td>2024148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox3-A</td>\n",
       "      <td>PB.8616</td>\n",
       "      <td>2058396</td>\n",
       "      <td>2065953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox4-A</td>\n",
       "      <td>g9720</td>\n",
       "      <td>2195412</td>\n",
       "      <td>2206712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox5-A</td>\n",
       "      <td>g9721</td>\n",
       "      <td>2351936</td>\n",
       "      <td>2354374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox6-A</td>\n",
       "      <td>PB.8617</td>\n",
       "      <td>2373415</td>\n",
       "      <td>2375678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox7-A</td>\n",
       "      <td>g9723</td>\n",
       "      <td>2565196</td>\n",
       "      <td>2594468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox8-A</td>\n",
       "      <td>g9724</td>\n",
       "      <td>2916314</td>\n",
       "      <td>2926445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>Hox10-A</td>\n",
       "      <td>g9725</td>\n",
       "      <td>2986021</td>\n",
       "      <td>2996225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             seqid gene_name        gene_id    start      end\n",
       "0   pseudochrom_56    Hox1-A        PB.8615  1927066  1936157\n",
       "1   pseudochrom_56            gap_PB.8615-0  1936158  1936159\n",
       "2   pseudochrom_56            gap_PB.8615-1  1936160  1936161\n",
       "3   pseudochrom_56            gap_PB.8615-2  1936162  1936163\n",
       "4   pseudochrom_56            gap_PB.8615-3  1936164  1936165\n",
       "5   pseudochrom_56    Hox2-A          g9718  1998922  2024148\n",
       "6   pseudochrom_56    Hox3-A        PB.8616  2058396  2065953\n",
       "7   pseudochrom_56    Hox4-A          g9720  2195412  2206712\n",
       "8   pseudochrom_56    Hox5-A          g9721  2351936  2354374\n",
       "9   pseudochrom_56    Hox6-A        PB.8617  2373415  2375678\n",
       "10  pseudochrom_56    Hox7-A          g9723  2565196  2594468\n",
       "11  pseudochrom_56    Hox8-A          g9724  2916314  2926445\n",
       "12  pseudochrom_56   Hox10-A          g9725  2986021  2996225"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | hide\n",
    "import nbdev\n",
    "\n",
    "nbdev.nbdev_export()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
